source/schemas/types.txt

```diff
@@ -509,3 +509,80 @@ with the ``title`` field set to the embedded object's type name:
             }
       }
    }
+
+.. _geospatial:
+
+Geospatial Data
+---------------
+
+Geospatial data describes points and other data on the earth's surface. App 
+Services supports defining GeoJSON Point data in a schema. 
+
+GeoJSON Point
+~~~~~~~~~~~~~
+
+A GeoJSON Point is a single location on the Earth's surface, described by its 
+longitude and latitude. To ensure wide compatibility, App Services stores these 
+points in accordance to the `GeoJSON spec <https://datatracker.ietf.org/doc/html/rfc7946>`_.
+To store a GeoPoint in Atlas, you must follow these rules:
+
+- It must be of type ``object``
+- It must be embedded within another type
+- It must contain the following fields:
+
+  - ``type``, which is always set to "Point"
+  - ``coordinates``, of type ``array``. The ``coordinates`` array contains 
+    exactly 2 double values. The first double is the longitude of the point and 
+    the second is the latitude.
+
+- Additional properties are allowed.
+ 
+The following is the schema of an object where the "location" property is 
+a GeoPoint:
+
+.. code-block:: json
+
+   {
+     "title": "MyObject",
+     "properties": {
+       "location": {
+         "bsonType": "object",
+         "required": [ "type" ],
+         "properties": {
+           "type": {
+             "bsonType": "string",
+             "const": "Point"
+           },
+           "coordinates": {
+             "bsonType": "array",
+             "items": {
+               "bsonType": "double",
+                "minItems": 2,
+                "maxItems": 2
+             }
+           }
+         }
+       },
+       "additional_location_property": {
+             "bsonType": "string"
+       },
+     }
+   }
+
```

Comment:
I think there are more things that we should mention in the sync-specific docs here; especially because it seems unclear to me at the moment whether future geospatial support will be rolled out in Realm at the same time as Atlas Device Sync.
1. `geoWithin` is the only operator supported in querying in Atlas Device Sync
2. GeoJSON Points are the _only_ form of geospatial data that can be queried in Atlas Device Sync. Note that this does not mean that you can't use the `geoWithin` operator against the supported geospatial types in Realm (eg. `location geoWithin GeoCircle([-73.981209, 40.725055], 0.00050464271`). It's just that the field name being operated on in the geoWithin operator (the LHS of the geoWithin operator) can only correspond to a GeoJSON Point property.
3. The sync server will perform GeoJSON Point validation for changes to **_any_** property that matches the JSON schema embedded object structure noted above (note that this is not just a GeoJSON Point property that is queryable, any property in the JSON schema that matches the embedded object structure above is subject to this!). The checks will specifically ensure that: (1) the "type" field can only be set to the string "Point" (case-sensitive), and (2) the "coordinates" field can only be set to an array of doubles. The first element in the "coordinates" array (longitude) must fall in the range of [-180, 180], and the second element (latitude) must fall in the range of [-90, 90].
  i. If a **device** makes a write that violates the above constraints, then it will be reverted via a [compensating write](https://www.mongodb.com/docs/atlas/app-services/sync/error-handling/errors/#std-label-flexible-sync-errors). 
  ii. If a **write to MongoDB** violates the above constraints, then the offending document will become [unsyncable](https://www.mongodb.com/docs/atlas/app-services/mongodb/internal-database-usage/#unsynced-documents). 
4. There is now a new option in the "BsonType" Select dropdown in the Rich Schema editor in the Schema Page UI that will automatically fill in the necessary requirements for marking a property as a GeoJSON Point. See [here](https://github.com/10gen/baas-ui/pull/3413) for what that looks like.

Reply:
@jonatlee - confused on point 2. Are you saying any embedded object that "looks like" a GeoPoint is validated? And if so, what does "look like" mean, because if it doesn't meet all the requirements of a GeoPoint, does it get skipped by the validator, or fail validation? How do we draw the line there?

Reply:
Yes, any property that is considered to be a GeoPoint according to the server's understanding of what a GeoPoint is is subject to the validation. If it does not meet all the requirements for specifying a GeoPoint, then no GeoPoint validation will come into play for that property. The line is drawn in that the following minimum requirements must be met in order for it to be considered a GeoPoint:
1. The property is an embedded object
2. The property contains a required field, "type", that is a string
3. The property contains a field, "coordinates", that must be an array of doubles

For example, the "location" field in the example below would fulfill these minimum requirements and be considered a GeoPoint by the server
```
{
    "title": "myTable",
    "required": ["_id"]
    "properties": {
        "_id": { "bsonType": "objectId" },
        "location": {
            "bsonType": "object",
            "required": [ "type" ],
            "properties": {
                "type": {
                    "bsonType": "string"
                },
                "coordinates": {
                    "bsonType": "array",
                    "items": {
                        "bsonType": "double"
                    }
                }
            }
        }
    }
}
```

Reply:
@jonatlee My question wasn't clear. I understand what is required to pass validation, but what I don't understand is this:
> any property that is considered to be a GeoPoint according to the server's understanding of what a GeoPoint is is subject to the validation

So the server understands a Geopoint has an embedded object with the above requirements, so in order for it to identify a Geopoint, it has already validated that's what it is? 

Or is it: "I see an embedded object with a 'type' and 'coordinates' properties, but 'type' isn't set to 'Point' (and/or 'coordinates isn't an array of >=2 doubles)"?

Reply:
The JSON schema requirements for a GeoPoint exists to answer the question of "should we run the GeoPoint validation checks against this property" whenever a write is made by a mobile client or directly to MongoDB. 

On each write, we look at the properties in the updated document. Any property in the updated document is cross-referenced with its JSON schema structure to see if it is a GeoPoint, and if it is one, then we will perform the GeoPoint validation checks on the updated state of the document. This is what allows us to ensure things like "type" only being allowed to be set to "Point" and "coordinates" only containing a list of doubles with valid longitude/latitude values outside of the JSON schema definition. 

Let me know if that makes sense. Also happy to chat over Slack about this if that's easier